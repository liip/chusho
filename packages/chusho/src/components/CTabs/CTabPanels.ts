import { VNodeData } from 'vue/types/umd';
import { defineComponent, createElement } from '@vue/composition-api';
import TabsMixin from './mixin';

interface TabPanelsProps {
  bare?: boolean;
}

export default defineComponent<TabPanelsProps>({
  name: 'CTabPanels',

  mixins: [TabsMixin],

  props: {},

  setup(props, { attrs, parent, slots }) {
    const tabsConfig = parent!.$chusho?.options?.components?.tabs;

    return () => {
      const componentData: VNodeData = {
        attrs,
        class: props.bare ? null : tabsConfig?.tabPanelsClass,
      };

      return createElement(
        'div',
        componentData,
        slots.default && slots.default()
      );
    };
  },
});
