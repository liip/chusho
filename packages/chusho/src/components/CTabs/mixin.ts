export default {
  props: {
    /**
     * Disable class inheritance from the component config.
     */
    bare: {
      type: Boolean,
      default: false,
    },
  },
};
