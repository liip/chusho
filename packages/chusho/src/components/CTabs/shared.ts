export const props = {
  /**
   * Disable class inheritance from the component config.
   */
  bare: {
    type: Boolean,
    default: false,
  },
};
