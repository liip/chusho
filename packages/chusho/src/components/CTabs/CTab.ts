import { defineComponent, h, inject, mergeProps, PropType } from 'vue';

import { TabsSymbol, UseTabs } from './CTabs';
import { props as sharedProps } from './shared';
import { ClassGenerator, DollarChusho, VueClassBinding } from '../../types';

export default defineComponent({
  name: 'CTab',

  inheritAttrs: false,

  props: {
    ...sharedProps,

    /**
     * The id of the Tab this button should control.
     */
    target: {
      type: [Number, String],
      required: true,
    },
    /**
     * Generate class based on the current component state. It should return a valid Vue “class” syntax (object, array or string), see [Vue class documentation](https://vuejs.org/v2/guide/class-and-style.html).
     *
     * For example: `(active) => ({ 'btn': true, 'btn--active': active })`
     */
    classGenerator: {
      type: Function as PropType<ClassGenerator>,
      default: null,
    },
  },

  setup(props, { attrs, slots }) {
    const tabs = inject(TabsSymbol) as UseTabs;

    return () => {
      const tabsConfig = inject<DollarChusho | null>('$chusho', null)?.options
        ?.components?.tabs;
      const isActive = props.target === tabs.currentTab.value;
      const elementProps = {
        type: 'button',
        id: `chusho-tabs-${tabs.uuid}-tab-${props.target}`,
        role: 'tab',
        'aria-selected': `${isActive}`,
        'aria-controls': `chusho-tabs-${tabs.uuid}-tabpanel-${props.target}`,
        tabindex: isActive ? '0' : '-1',
        class: <VueClassBinding[]>[],
        onClick() {
          if (!props.target) return;
          tabs.setCurrentTab(props.target);
        },
      };

      if (!props.bare && tabsConfig?.tabClass) {
        if (typeof tabsConfig.tabClass === 'function') {
          elementProps.class.push(tabsConfig.tabClass(isActive));
        } else {
          elementProps.class.push(tabsConfig.tabClass);
        }
      }

      if (props.classGenerator) {
        elementProps.class.push(props.classGenerator(isActive));
      }

      return h('button', mergeProps(attrs, elementProps), slots);
    };
  },
});
