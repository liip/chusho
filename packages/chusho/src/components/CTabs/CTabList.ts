import { defineComponent, h, inject, ref, nextTick, mergeProps } from 'vue';

import { TabsSymbol, UseTabs } from './CTabs';
import { props as sharedProps } from './shared';
import { getSiblingIndexByArrowKey } from '../../utils/keyboard';
import { DollarChusho } from '../../types';

export default defineComponent({
  name: 'CTabList',

  inheritAttrs: false,

  props: {
    ...sharedProps,
  },

  setup(props, { attrs, slots }) {
    const chushoOptions = inject<DollarChusho | null>('$chusho', null)?.options;
    const tabs = inject(TabsSymbol) as UseTabs;
    const tabList = ref<null | Element>(null);

    function handleNavigation(e: KeyboardEvent): void {
      if (!tabs.currentTab.value) return;

      const activeTabIndex = tabs.tabs.value.indexOf(tabs.currentTab.value);
      const activeTabId = tabs.tabs.value[activeTabIndex];
      const rtl = chushoOptions?.rtl;
      const newIndex = getSiblingIndexByArrowKey(
        tabs.tabs.value,
        activeTabId,
        e.key,
        rtl && rtl()
      );

      if (newIndex !== undefined) {
        e.preventDefault();

        tabs.setCurrentTab(tabs.tabs.value[newIndex]);

        nextTick(() => {
          if (tabList.value) {
            const activeTab: HTMLElement | null = tabList.value.querySelector(
              '[aria-selected="true"]'
            );
            if (activeTab) {
              activeTab.focus();
            }
          }
        });
      }
    }

    return () => {
      const tabsConfig = chushoOptions?.components?.tabs;
      const elementProps: Record<string, unknown> = {
        role: 'tablist',
        onKeydown: handleNavigation,
        ref: tabList,
      };

      if (!props.bare && tabsConfig?.tabListClass) {
        elementProps.class = tabsConfig.tabListClass;
      }

      return h('div', mergeProps(attrs, elementProps), slots);
    };
  },
});
