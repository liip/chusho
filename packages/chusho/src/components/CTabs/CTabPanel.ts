import { defineComponent, h, inject, mergeProps, PropType } from 'vue';
import { SelectedItemId } from '../../composables/useSelected';

import { DollarChusho } from '../../types';
import { generateConfigClass } from '../../utils/components';
import componentMixin from '../mixins/componentMixin';
import { TabsSymbol } from './CTabs';

export default defineComponent({
  name: 'CTabPanel',

  mixins: [componentMixin],

  inheritAttrs: false,

  props: {
    /**
     * A unique ID to target the panel with CTab.
     */
    id: {
      type: [String, Number] as PropType<SelectedItemId>,
      required: true,
    },
  },

  setup(props) {
    const tabs = inject(TabsSymbol);

    tabs?.addItem(props.id);

    // Set the first tab as active if none has been defined
    if (tabs?.items.value.length === 1 && !tabs?.selectedItemId.value) {
      tabs.setSelectedItem(props.id);
    }

    return {
      tabs,
    };
  },

  beforeUnmount() {
    this.tabs?.removeItem(this.id);
  },

  render() {
    if (!this.tabs) return;

    const tabPanelConfig = inject<DollarChusho | null>('$chusho', null)?.options
      ?.components?.tabPanel;
    const isActive = this.id === this.tabs.selectedItemId.value;

    if (!isActive) return null;

    const elementProps = {
      id: `${this.tabs.uuid}-tabpanel-${this.id}`,
      role: 'tabpanel',
      'aria-labelledby': `${this.tabs.uuid}-tab-${this.id}`,
      tabindex: '0',
      ...generateConfigClass(tabPanelConfig?.class, {
        ...this.$props,
        active: isActive,
      }),
    };

    return h('div', mergeProps(this.$attrs, elementProps), this.$slots);
  },
});
