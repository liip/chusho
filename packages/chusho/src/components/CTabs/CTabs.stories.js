import CTabs from './CTabs';
import CTabList from './CTabList';
import CTab from './CTab';
import CTabPanels from './CTabPanels';
import CTabPanel from './CTabPanel';

export default {
  title: 'Tabs',
  component: CTabs,
  subcomponents: { CTabList, CTab, CTabPanels, CTabPanel },
};

export const Default = () => ({
  components: { CTabs, CTabList, CTab, CTabPanels, CTabPanel },
  template: `
    <CTabs>
      <CTabList aria-label="Example of tabs">
        <CTab target="1">Tab 1</CTab>
        <CTab target="2">Tab 2</CTab>
        <CTab target="3">Tab 3</CTab>
      </CTabList>

      <CTabPanels>
        <CTabPanel id="1">
          Lorem ipsum dolor sit amet consectetur adipisicing elit. Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab.
        </CTabPanel>
        <CTabPanel id="2">
          Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab, libero quo maiores sapiente doloribus nihil commodi eaque accusantium praesentium! Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
        </CTabPanel>
        <CTabPanel id="3">
          Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
        </CTabPanel>
      </CTabPanels>
    </CTabs>
  `,
});

export const Controlled = () => ({
  components: { CTabs, CTabList, CTab, CTabPanels, CTabPanel },
  data() {
    return {
      currentTab: '2',
    };
  },
  template: `
    <CTabs v-model="currentTab">
      <CTabList aria-label="Example of controlled tabs">
        <CTab target="1">Tab 1</CTab>
        <CTab target="2">Tab 2</CTab>
        <CTab target="3">Tab 3</CTab>
      </CTabList>

      <CTabPanels>
        <CTabPanel id="1">
          Lorem ipsum dolor sit amet consectetur adipisicing elit. Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab.
        </CTabPanel>
        <CTabPanel id="2">
          Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab, libero quo maiores sapiente doloribus nihil commodi eaque accusantium praesentium! Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
        </CTabPanel>
        <CTabPanel id="3">
          Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
        </CTabPanel>
      </CTabPanels>
    </CTabs>
  `,
});

Controlled.story = {
  parameters: {
    docs: {
      storyDescription:
        'You can optionally control wich Tab is active with the `v-model` directive.',
    },
  },
};

export const OverrideStyle = () => ({
  components: { CTabs, CTabList, CTab, CTabPanels, CTabPanel },

  methods: {
    getTabClass(active) {
      return {
        'text-red-700': !active,
        'text-white font-semibold bg-red-700': active,
      };
    },
  },

  template: `
    <CTabs defaultTab="3">
      <CTabList aria-label="Example of tabs with different style" class="flex justify-center mb-3" bare>
        <CTab target="1" :classGenerator="getTabClass" class="py-2 px-5 rounded-full" bare>Tab 1</CTab>
        <CTab target="2" :classGenerator="getTabClass" class="py-2 px-5 rounded-full" bare>Tab 2</CTab>
        <CTab target="3" :classGenerator="getTabClass" class="py-2 px-5 rounded-full" bare>Tab 3</CTab>
      </CTabList>

      <CTabPanels>
        <CTabPanel id="1" bare>
          Lorem ipsum dolor sit amet consectetur adipisicing elit. Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab.
        </CTabPanel>
        <CTabPanel id="2" bare>
          Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab, libero quo maiores sapiente doloribus nihil commodi eaque accusantium praesentium! Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
        </CTabPanel>
        <CTabPanel id="3" bare>
          Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
        </CTabPanel>
      </CTabPanels>
    </CTabs>
  `,
});

OverrideStyle.story = {
  parameters: {
    docs: {
      storyDescription:
        'Use the `bare` prop on any component to remove the class inherited from the config and create different styles from scratch.',
    },
  },
};
