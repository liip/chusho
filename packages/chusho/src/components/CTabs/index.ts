import CTabs from './CTabs';
import CTabList from './CTabList';
import CTab from './CTab';
import CTabPanels from './CTabPanels';
import CTabPanel from './CTabPanel';

export { CTabs, CTabList, CTab, CTabPanels, CTabPanel };
