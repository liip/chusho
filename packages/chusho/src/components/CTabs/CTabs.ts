import {
  provide,
  computed,
  InjectionKey,
  defineComponent,
  h,
  watchEffect,
  Ref,
  reactive,
  PropType,
  inject,
  mergeProps,
} from 'vue';
import uuid from '../../utils/uuid';

import { DollarChusho } from '../../types';
import { generateConfigClass } from '../../utils/components';
import componentMixin from '../mixin';

export const TabsSymbol: InjectionKey<UseTabs> = Symbol();

export declare type TabId = string | number;

interface TabsState {
  currentTab?: TabId;
  tabs: TabId[];
}

export interface UseTabs {
  uuid: string;
  currentTab: Readonly<Ref<TabId | undefined>>;
  tabs: Readonly<Ref<readonly TabId[]>>;
  setCurrentTab: (id: TabId) => void;
  registerTab: (id: TabId) => void;
}

export default defineComponent({
  name: 'CTabs',

  mixins: [componentMixin],

  inheritAttrs: false,

  props: {
    /**
     * Optionally bind the Tabs state with the parent component.
     */
    modelValue: {
      type: [Number, String] as PropType<TabId>,
      default: null,
    },
    /**
     * The id of the Tab to display by default. This is ignored if `v-model` is used.
     *
     * Defaults to the first tab.
     */
    defaultTab: {
      type: [Number, String] as PropType<TabId>,
      default: null,
    },
  },

  emits: ['update:modelValue'],

  setup(props, { emit }) {
    const state = reactive<TabsState>({
      currentTab: props.modelValue || props.defaultTab,
      tabs: [],
    });

    function setCurrentTab(id: TabId) {
      state.currentTab = id;
      // Update potential parent v-model value
      if (['string', 'number'].includes(typeof props.modelValue)) {
        emit('update:modelValue', id);
      }
    }

    function registerTab(id: TabId) {
      state.tabs.push(id);

      if (state.tabs.length === 1 && !state.currentTab) {
        state.currentTab = id;
      }
    }

    // Provide api to sub-components
    const api: UseTabs = {
      uuid: uuid(),
      currentTab: computed(() => state.currentTab),
      tabs: computed(() => state.tabs),
      setCurrentTab,
      registerTab,
    };
    provide(TabsSymbol, api);

    // Watch potential parent v-model value changes and update state accordingly
    watchEffect(() => {
      if (['string', 'number'].includes(typeof props.modelValue)) {
        state.currentTab = props.modelValue;
      }
    });
  },

  render() {
    const tabsConfig = inject<DollarChusho | null>('$chusho', null)?.options
      ?.components?.tabs;
    const elementProps: Record<string, unknown> = {
      ...generateConfigClass(tabsConfig?.class, this.$props),
    };

    return h('div', mergeProps(this.$attrs, elementProps), this.$slots);
  },
});
