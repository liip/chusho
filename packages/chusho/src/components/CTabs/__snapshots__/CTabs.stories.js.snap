// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Tabs Controlled 1`] = `
<div
  class="tabs"
>
  <div
    aria-label="Example of controlled tabs"
    class="flex mx-4"
    role="tablist"
  >
    <button
      aria-controls="chusho-tabs-4-tabpanel-1"
      aria-selected="false"
      class="inline-block py-2 px-5 border-b-2 border-transparent text-gray-700"
      id="chusho-tabs-4-tab-1"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Tab 1
    </button>
     
    <button
      aria-controls="chusho-tabs-4-tabpanel-2"
      aria-selected="true"
      class="inline-block py-2 px-5 border-b-2 border-transparent text-blue-800 bg-blue-100 border-blue-300 rounded-t"
      id="chusho-tabs-4-tab-2"
      role="tab"
      tabindex="0"
      type="button"
    >
      Tab 2
    </button>
     
    <button
      aria-controls="chusho-tabs-4-tabpanel-3"
      aria-selected="false"
      class="inline-block py-2 px-5 border-b-2 border-transparent text-gray-700"
      id="chusho-tabs-4-tab-3"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Tab 3
    </button>
  </div>
   
  <div
    class="py-3 px-4 bg-gray-200 rounded"
  >
    <!---->
     
    <div
      aria-labelledby="chusho-tabs-4-tab-2"
      class="tabpanel"
      id="chusho-tabs-4-tabpanel-2"
      role="tabpanel"
      tabindex="0"
    >
      
          Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab, libero quo maiores sapiente doloribus nihil commodi eaque accusantium praesentium! Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
        
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`Storyshots Tabs Default 1`] = `
<div
  class="tabs"
>
  <div
    aria-label="Example of tabs"
    class="flex mx-4"
    role="tablist"
  >
    <button
      aria-controls="chusho-tabs-3-tabpanel-1"
      aria-selected="false"
      class="inline-block py-2 px-5 border-b-2 border-transparent text-gray-700"
      id="chusho-tabs-3-tab-1"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Tab 1
    </button>
     
    <button
      aria-controls="chusho-tabs-3-tabpanel-2"
      aria-selected="false"
      class="inline-block py-2 px-5 border-b-2 border-transparent text-gray-700"
      id="chusho-tabs-3-tab-2"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Tab 2
    </button>
     
    <button
      aria-controls="chusho-tabs-3-tabpanel-3"
      aria-selected="false"
      class="inline-block py-2 px-5 border-b-2 border-transparent text-gray-700"
      id="chusho-tabs-3-tab-3"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Tab 3
    </button>
  </div>
   
  <div
    class="py-3 px-4 bg-gray-200 rounded"
  >
    <div
      aria-labelledby="chusho-tabs-3-tab-1"
      class="tabpanel"
      id="chusho-tabs-3-tabpanel-1"
      role="tabpanel"
      tabindex="0"
    >
      
          Lorem ipsum dolor sit amet consectetur adipisicing elit. Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis fuga cupiditate suscipit blanditiis, aliquid minima harum molestias pariatur tempora ab.
        
    </div>
     
    <!---->
     
    <!---->
  </div>
</div>
`;

exports[`Storyshots Tabs Override Style 1`] = `
<div
  class="tabs"
>
  <div
    aria-label="Example of tabs with different style"
    class="flex justify-center mb-3"
    role="tablist"
  >
    <button
      aria-controls="chusho-tabs-5-tabpanel-1"
      aria-selected="false"
      class="py-2 px-5 rounded-full text-red-700"
      id="chusho-tabs-5-tab-1"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Tab 1
    </button>
     
    <button
      aria-controls="chusho-tabs-5-tabpanel-2"
      aria-selected="false"
      class="py-2 px-5 rounded-full text-red-700"
      id="chusho-tabs-5-tab-2"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Tab 2
    </button>
     
    <button
      aria-controls="chusho-tabs-5-tabpanel-3"
      aria-selected="true"
      class="py-2 px-5 rounded-full text-white font-semibold bg-red-700"
      id="chusho-tabs-5-tab-3"
      role="tab"
      tabindex="0"
      type="button"
    >
      Tab 3
    </button>
  </div>
   
  <div
    class="py-3 px-4 bg-gray-200 rounded"
  >
    <!---->
     
    <!---->
     
    <div
      aria-labelledby="chusho-tabs-5-tab-3"
      class=""
      id="chusho-tabs-5-tabpanel-3"
      role="tabpanel"
      tabindex="0"
    >
      
          Laboriosam in, iste id nobis dolor excepturi dolore expedita vero quae. Nobis natus qui voluptate inventore molestias quisquam, consequuntur harum?
        
    </div>
  </div>
</div>
`;
